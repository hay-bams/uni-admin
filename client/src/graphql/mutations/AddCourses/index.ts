import { gql } from "@apollo/client";

export const ADD_COURSES = gql`
  mutation AddCourse($studentId: String!, $input: ID!) {
    addCourses(studentId: $studentId, input: $input) {
      id
    }
  }
`;