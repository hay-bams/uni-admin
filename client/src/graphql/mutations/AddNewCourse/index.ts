import { gql } from "@apollo/client";

export  const ADD_NEW_COURSE = gql`
  mutation AddNewCourse($input: NewCourseInput) {
    addNewCourse(input: $input) {
      id
    }
  }
`