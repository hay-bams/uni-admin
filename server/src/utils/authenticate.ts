import { ObjectID } from 'mongodb';
import jwt from 'jsonwebtoken'
import {  ICtx } from '../lib/types';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const Authenticated =  <T=any, K=any> (next: any) => async (
  root: T,
  args: K,
  ctx: ICtx
) => {
  const { db, req, res } = ctx;
  const token = req.get("X-CSRF-TOKEN");

  const id_cookie = req.signedCookies.admin;
  const decoded = jwt.verify(`${token}`, `${process.env.SECRET}`) as { admin: string }

  if(id_cookie !== decoded.admin) {
    throw new Error('This may be a CSRF attack, please sign in')
  }

  const admin = await db.users.findOne({
    _id: new ObjectID(id_cookie),
  });

  if (!admin) {
    res.clearCookie('admin');
    throw new Error('User not authenticated, please sign in first');
  }

  return next(root, args, ctx);
};
