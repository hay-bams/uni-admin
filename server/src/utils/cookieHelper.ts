import { Request, Response } from "express"
import { ObjectID } from "mongodb"

import { Database } from "../lib/types"

export const cookieOptions = {
  httpOnly: true,
  sameSite: true,
  signed: true,
  secure: process.env.NODE_ENV === 'development' ? false : true
}

export const cookie = {
  setCookie: (id: ObjectID | undefined, res: Response) => {
    if(id) {
      res.cookie('admin', id, {
        ...cookieOptions,
        maxAge: 1000 * 60 * 60 * 24 * 365
      })
    }
  },
  loginViaCookie: async (db: Database, req: Request, res: Response) => {
    const result = await db.users.findOne({
      _id: new ObjectID(req.signedCookies.admin),
    });
  
    if (!result) {
      res.clearCookie('admin');
      return null;
    }
  
    return result;
  }
}