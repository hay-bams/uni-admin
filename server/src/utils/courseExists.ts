import { ObjectID } from "mongodb";
import { Course, Database, Student } from "../lib/types";

export const curseExists = async (
  ids: string[],
  student: Student,
  db: Database
): Promise<Course | null> => {
  const courses = student.courses;
  const map = new Map();

  for (const course of courses) {
    map.set(course.toString(), course.toString());
  }

  for (const id of ids) {
    if (map.get(id)) {
      const course = await db.courses.findOne({
        _id: new ObjectID(map.get(id))
      })

      if(course) return course;
    }
  }
  return null;
};
