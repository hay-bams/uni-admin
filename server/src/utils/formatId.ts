import { ObjectID } from "mongodb"

export const formatId = (ids: string[]) => {
  const newIds = ids.map((id) => {
    return new ObjectID(id)
  })

  return newIds
}