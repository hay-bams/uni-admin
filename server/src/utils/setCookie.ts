import { Response } from "express"
import { ObjectID } from "mongodb"

export const cookieOptions = {
  httpOnly: true,
  sameSite: true,
  signed: true,
  secure: process.env.NODE_ENV === 'development' ? false : true
}

export const setCookie = (id: ObjectID | undefined, res: Response) => {
  if(id) {
    res.cookie('user', id, {
      ...cookieOptions,
      maxAge: 1000 * 60 * 60 * 24 * 365
    })
  }
}