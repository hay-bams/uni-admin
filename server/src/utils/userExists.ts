import { Database } from "../lib/types";

export const userExist = async (db: Database, username: string) => {
  const user = await db.users.findOne({
    username,
  });

  if (user) {
    return true;
  }

  return false;
};