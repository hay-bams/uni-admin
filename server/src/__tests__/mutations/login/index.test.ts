/* eslint-disable @typescript-eslint/no-non-null-assertion */
import bcrypt from 'bcryptjs';

import { USER_DATA } from '../../test_data';
import { createTestServer } from '../../helpers';
import { TEST_LOG_IN } from '../mutations';
import { cookie } from '../../../utils/cookieHelper';

jest.mock('bcryptjs');
const mockedBcrypt = bcrypt as jest.Mocked<typeof bcrypt>;
const bcryptCompare = mockedBcrypt.compare;

beforeEach(() => {
  bcryptCompare.mockReset();
});

describe('Admin Login Mutation', () => {
  test('should login admin', async () => {
    const { mutate } = createTestServer({
      db: {
        users: {
          findOne: jest.fn(() => USER_DATA[0]),
        },
      },
    });

    

    // bcryptCompare = jest.fn().mockReturnValue(true)
    // bcryptCompare.mockResolvedValue(true);
    jest.spyOn(bcrypt, 'compare').mockImplementation(jest.fn(() => true))

    jest.spyOn(cookie, 'setCookie').mockImplementation(jest.fn());

    const res = await mutate({
      mutation: TEST_LOG_IN,
      variables: {
        input: {
          username: 'haybams',
          password: 'password',
          withCookie: false,
        },
      },
    });

    expect(res.data.login).toHaveProperty('id');
    expect(res.data.login).toHaveProperty('username');
    expect(res.data.login).toHaveProperty('madeRequest');
    expect(res.data.login.madeRequest).toBe(true);
    expect(res).toMatchSnapshot();
  });

  test('should login via cookie', async () => {
    const { mutate } = createTestServer({
      db: {
        users: {
          findOne: jest.fn(() => USER_DATA[0]),
        },
      },
    });

    jest.spyOn(cookie, 'loginViaCookie').mockImplementation(jest.fn()); 

    const res = await mutate({
      mutation: TEST_LOG_IN,
      variables: {
        input: {
          username: USER_DATA[0].username,
          password: 'passwordu',
          withCookie: true,
        },
      },
    });

    expect(res.data.login).toHaveProperty('id');
    expect(res.data.login).toHaveProperty('username');
    expect(res.data.login).toHaveProperty('madeRequest');
    expect(res.data.login.madeRequest).toBe(true);
  });

  test('should return an error if user is not found', async () => {
    const { mutate } = createTestServer({
      db: {
        users: {
          findOne: jest.fn(() => null),
        },
      },
    });

    const res = await mutate({
      mutation: TEST_LOG_IN,
      variables: {
        input: {
          username: 'haybams',
          password: 'password',
          withCookie: false,
        },
      },
    });

    expect(res.errors?.length).toBeGreaterThan(0);
   expect(res.errors![0]).toHaveProperty('message');
    expect(res).toMatchSnapshot();
  });

  test('should throw an if pasword is wrong', async () => {
    const { mutate } = createTestServer({
      db: {
        users: {
          findOne: jest.fn(() => USER_DATA[0].username),
        },
      },
    });

    jest.spyOn(bcrypt, 'compare').mockImplementation(jest.fn(() => false))
    // bcryptCompare.mockResolvedValue(false);
    // bcryptCompare = jest.fn().mockReturnValue(true)

    const res = await mutate({
      mutation: TEST_LOG_IN,
      variables: {
        input: {
          username: 'haybams',
          password: 'wrongPassword',
          withCookie: false,
        },
      },
    });

    expect(res.errors?.length).toBeGreaterThan(0);
    expect(res.errors![0]).toHaveProperty('message');
    expect(res).toMatchSnapshot();
  });
});
