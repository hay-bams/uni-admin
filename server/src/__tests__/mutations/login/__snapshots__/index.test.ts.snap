// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Admin Login Mutation should login admin 1`] = `
Object {
  "data": Object {
    "login": Object {
      "id": "5fe09161251c8b0a9565108d",
      "madeRequest": true,
      "username": "admin",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Admin Login Mutation should return an error if user is not found 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Something went wrong: Error: User does not exist, Please confirm your username is correct],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Admin Login Mutation should throw an if pasword is wrong 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Something went wrong: Error: Wrong Password, Please Try again with the correct password],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
